### PR TITLE
chore(typings): added type signatures for concat/exhaust/merge/switch

### PR DESCRIPTION
**Pull request by david-driscoll**
also adds signatures for the "All", "Map", and "MapTo" variants.

### CODE REVIEW

File: src/operator/mergeMapTo.ts [hunk 1]
```diff
@@ -6,35 +6,43 @@ import {Subscription} from '../Subscription';
...
   private hasCompleted: boolean = false;
   private buffer: Observable<any>[] = [];
   private active: number = 0;
   protected index: number = 0;
 
-  constructor(destination: Subscriber<R2>,
-              private ish: Observable<R> | Promise<R>,
-              private resultSelector?: (outerValue: T, innerValue: R, outerIndex: number, innerIndex: number) => R2,
+  constructor(destination: Subscriber<R>,
+              private ish: Observable<I> | Promise<I>,
```
Comment (kwonoj) [inline src/operator/mergeMapTo.ts]: can this be `ObservableOrPromise<T>`?

Comment (david-driscoll) [inline src/operator/mergeMapTo.ts]: it can, updated.

**Comment by kwonoj** [comment]
Change looks good to me. Will try to check in after https://github.com/ReactiveX/RxJS/pull/1352.

**Comment by david-driscoll** [comment]
this may need a rebase, so I'll keep an eye on on #1352, and get that done asap

**Comment by kwonoj** [comment]
@david-driscoll seems it requires rebase as you've expected. :) In general I think change's ok, cc @masaeedu for visibility just in case.

File: src/operator/mergeMap.ts [hunk 2]
```diff
@@ -17,34 +17,41 @@ import {InnerSubscriber} from '../InnerSubscriber';
  * @returns {Observable} an Observable that emits the result of applying the transformation function to each item
  * emitted by the source Observable and merging the results of the Observables obtained from this transformation
  */
-export function mergeMap<T, R, R2>(project: (value: T, index: number) => Observable<R>,
-                                   resultSelector?: (outerValue: T, innerValue: R, outerIndex: number, innerIndex: number) => R2 | number,
-                                   concurrent: number = Number.POSITIVE_INFINITY): Observable<R2> {
+export function mergeMap<T, I, R>(project: (value: T, index: number) => ObservableInput<I>,
+                                  resultSelector?: (outerValue: T, innerValue: I, outerIndex: number, innerIndex: number) => R | number,
```
Comment (masaeedu) [inline src/operator/mergeMap.ts]: Grouping issue again; I think this should be `((outerValue: T, innerValue: I, outerIndex: number, innerIndex: number) => R) | number`, which makes `mergeMap` assignable to its signature.

**Comment by masaeedu** [comment]
@kwonoj Similar non-blocking quibbles as the last one. LGTM.

**Comment by kwonoj** [comment]
Cool, let me kick off this once it's rebased.

**Comment by david-driscoll** [comment]
@kwonoj @masaeedu rebased.

**Comment by kwonoj** [comment]
@david-driscoll seems there was some failure with type definition update?

**Comment by david-driscoll** [comment]
Yeah, missed that.  Updated now.

**Comment by kwonoj** [comment]
Merged with 1933774, thanks @david-driscoll !

**Pull request merged**